File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/UncheckedTasksRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that forbids unchecked tasks in completed plan documents.
 */
public enum UncheckedTasksRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "uncheckedTasks";

    @Override
    public String category() {
        return "uncheckedTasks";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final JsonNode catNode = manifest.get(CATEGORY);
        final String directory = catNode.get("parameters").get("directory").asText();
        final String uncheckedPattern = catNode.get("parameters").get("uncheckedTaskPattern").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final Path dirPath = root.resolve(directory);
        final Pattern pattern = Pattern.compile(uncheckedPattern);
        return Stream.of(dirPath)
                .filter(p -> HarnessCheckHelper.isSafeDirectory(root, p))
                .flatMap(p -> {
                    try {
                        return HarnessCheckHelper.safeFileOrWalk(root, p).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .filter(f -> f.getFileName().toString().endsWith(".md"))
                .filter(f -> !f.getFileName().toString().equals(".gitkeep"))
                .flatMap(file -> {
                    try {
                        final String text = HarnessCheckHelper.readFile(root, file);
                        return pattern.matcher(text).find()
                                ? Stream.of(Finding.of(severity, CATEGORY, "completed plan has unchecked tasks: " + root.relativize(file)))
                                : Stream.empty();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/fs/UncheckedTasksRule.kt
package com.ririnto.sinon.harness.rules.fs

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.name
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that forbids unchecked tasks in completed plans.
 */
object UncheckedTasksRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "uncheckedTasks"

    override fun validate(ctx: RuleContext): Collection<Finding> {
        val parametersObj = (ctx.manifest.categoryObject(category) ?: return emptyList()).get("parameters")?.jsonObject ?: return emptyList()
        val directory = ctx.root / JsonAccess.stringFromObject(parametersObj, "directory")
        if (!directory.isDirectory()) {
            return emptyList()
        }
        return buildList {
            ctx
                .walkSafe(directory)
                .paths
                .filter { file -> file.name.endsWith(".md") }
                .filter { file -> JsonAccess.stringFromObject(parametersObj, "uncheckedTaskPattern").toRegex().containsMatchIn(file.readText()) }
                .forEach { file ->
                    add(
                        Finding(
                            ctx.manifest.severityOf(category),
                            category,
                            ctx.manifest.stringValue(category, "default").takeIf { message -> message.isNotEmpty() }
                                ?: "completed plan has unchecked tasks: ${file.relativeTo(ctx.root)}",
                        ),
                    )
                }
        }
    }
}
